### ISSUE TITLE
self is not defined

### ISSUE THREAD

**Issue by samkelleher** [mozilla/localForage#469]
See https://github.com/mozilla/localForage/blob/master/src/localforage.js#L424

```
call(typeof window !== 'undefined' ? window : self);
```

The `self` is undefined. Although I am uncertain what it is supposed to be. It looks like the passed in variable isn't used. The self might have been added in a pull request adding service worker support.

This error prevents localforage being upstream with webpack.

https://github.com/mozilla/localForage/commit/66ed3aeb261886c60f7805a1ccdbbb59d675eae0

**Comment by thgreasi** [comment]
Yes, this was added to support service & web workers.
> This error prevents localforage being upstream with webpack.

What exactly is your error?
Because we are using webpack to bundle localforage's sources in v1.3 and never got any error!?!?

**Comment by samkelleher** [comment]
I see that webpack is used to bundle localforage itself, but then if you consume the localforage dist into another project that is also being bundled with webpack, then webpack throws a wobbly because it can't resolve `self`.

For example, in a project where localforage is installed via npm, import it your file, then bundle this. Webpack will go off and load the dist output.

```
import localforage from 'localforage';
```

**Comment by thgreasi** [comment]
As said in #470,
> We have webpack [repacking](https://github.com/mozilla/localForage/blob/eb9d5b1a9f7416401a620fd6c307d45daae1f690/Gruntfile.js#L212-L219) [tests](https://github.com/mozilla/localForage/blob/master/test/test.webpack.html)

What exactly is your error and what configuration do you use?

**Comment by samkelleher** [comment]
This is the exact error thrown by webpack when bundling it in with `myproject` that has a file with the import as mentioned above.

I'll look at the test (just grabbing 🍜lunch) and figure out why the test passes but this fails.

Cheers

```
/myproject/node_modules/localforage/dist/localforage.js:1161
        typeof window !== 'undefined' ? window : self
                                                 ^

ReferenceError: self is not defined
    at Object.exports.__esModule (/myproject/node_modules/localforage/dist/localforage.js:1161:50)
    at __webpack_require__ (/myproject/node_modules/localforage/dist/localforage.js:719:30)
    at exports.__esModule (/myproject/node_modules/localforage/dist/localforage.js:739:18)
    at /myproject/node_modules/localforage/dist/localforage.js:742:10
    at webpackUniversalModuleDefinition (/myproject/node_modules/localforage/dist/localforage.js:692:20)
    at Object.<anonymous> (/myproject/node_modules/localforage/dist/localforage.js:699:3)
```

**Comment by mgcrea** [comment]
I'm encountering the same error using `import localforage from 'localforage';` + babel6. Did you find some solution?

**Comment by thgreasi** [comment]
Is there an error trace available? Are you targeting the `/src/` files?

**Comment by mgcrea** [comment]
File is compiled (sucessfully) by babel:

```
import extend from 'lodash/object/extend';
import clone from 'lodash/lang/clone';
import localforage from 'localforage';
const log = console.log.bind(console);
log(localforage);
export default class FooBar {
}

```

- With `.babelrc`

```
{
  "presets": ["es2015", "stage-0"]
}
```

- But in my mocha unit tests:

```

import chai from 'chai';
const expect = chai.expect;
import FooBar from './../foobar.js';

describe('FooBar', () => {
  describe('#constructor()', () => {
    it('should properly merge options', () => {
      const analytics = new FooBar({foo: 'bar'});
      // assert.equal(analytics.opts, {foo: 'bar'});
      expect(analytics.opts).to.be.a('object');
      expect(analytics.opts.foo).to.eql('bar');
    });
  });
});
```

It fails right away with:

```
ReferenceError: self is not defined
    at Object.exports.__esModule (/Users/olivier/Projects/web-analytics/node_modules/localforage/dist/localforage.js:1157:51)
    at __webpack_require__ (/Users/olivier/Projects/web-analytics/node_modules/localforage/dist/localforage.js:719:30)
    at exports.__esModule (/Users/olivier/Projects/web-analytics/node_modules/localforage/dist/localforage.js:739:18)
    at /Users/olivier/Projects/web-analytics/node_modules/localforage/dist/localforage.js:742:10
    at webpackUniversalModuleDefinition (/Users/olivier/Projects/web-analytics/node_modules/localforage/dist/localforage.js:692:20)
    at Object.<anonymous> (/Users/olivier/Projects/web-analytics/node_modules/localforage/dist/localforage.js:699:3)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at Object.<anonymous> (/Users/olivier/Projects/web-analytics/dist/analytics.js:19:20)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at Object.<anonymous> (/Users/olivier/Projects/web-analytics/dist/test/analytics.spec.js:7:18)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at /Users/olivier/Projects/web-analytics/node_modules/mocha/lib/mocha.js:216:27
    at Array.forEach (native)
    at Mocha.loadFiles (/Users/olivier/Projects/web-analytics/node_modules/mocha/lib/mocha.js:213:14)
    at Mocha.run (/Users/olivier/Projects/web-analytics/node_modules/mocha/lib/mocha.js:453:10)
    at loadAndRun (/Users/olivier/Projects/web-analytics/node_modules/mocha/bin/_mocha:349:22)
    at rerun (/Users/olivier/Projects/web-analytics/node_modules/mocha/bin/_mocha:376:5)
    at /Users/olivier/Projects/web-analytics/node_modules/mocha/bin/_mocha:384:7
    at StatWatcher.<anonymous> (/Users/olivier/Projects/web-analytics/node_modules/mocha/lib/utils.js:173:9)
    at emitTwo (events.js:87:13)
    at StatWatcher.emit (events.js:172:7)
    at StatWatcher._handle.onchange (fs.js:1295:10)
```

**Comment by thgreasi** [comment]
I'm still not able to reproduce this... and the integration tests seem to
be passing !?!?
Can you test of #470 fixes your problem?


-- 
*Thodoris Greasidis*
Computer, Networking &
Communications Engineer

**Comment by mgcrea** [comment]
Juste merged the changed and rebuild. I do confirm that #470 does fix it.

**Comment by montogeek** [comment]
@thgreasi Hi!
I am getting this same error with:
```
"localforage": "1.4.0"
"babel-cli": "^6.6.5",
"babel-core": "^6.7.0"
"mocha": "^2.4.5"
```

The command is `./node_modules/.bin/mocha --recursive --compilers js:babel-core/register`
My .babelrc is:
```
{
  "presets": ["es2015"],
  "plugins": ["transform-flow-strip-types", "transform-object-rest-spread", "transform-class-properties"]
}
```

Stacktrace:
```
/Users/montogeek/project/node_modules/localforage/dist/localforage.js:1162
  })(typeof window !== 'undefined' ? window : self);
                                              ^

ReferenceError: self is not defined
    at Object.exports.__esModule (/Users/montogeek/project/node_modules/localforage/dist/localforage.js:1162:46)
    at __webpack_require__ (/Users/montogeek/project/node_modules/localforage/dist/localforage.js:719:30)
    at exports.__esModule (/Users/montogeek/project/node_modules/localforage/dist/localforage.js:739:18)
    at /Users/montogeek/project/node_modules/localforage/dist/localforage.js:742:10
    at webpackUniversalModuleDefinition (/Users/montogeek/project/node_modules/localforage/dist/localforage.js:692:20)
    at Object.<anonymous> (/Users/montogeek/project/node_modules/localforage/dist/localforage.js:699:3)
    at Module._compile (module.js:398:26)
    at Module._extensions..js (module.js:405:10)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/montogeek/project/node_modules/babel-register/lib/node.js:134:7)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (auth.js:3:1)
    at Module._compile (module.js:398:26)
    at loader (/Users/montogeek/project/node_modules/babel-register/lib/node.js:126:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/montogeek/project/node_modules/babel-register/lib/node.js:136:7)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (index.js:4:1)
    at Module._compile (module.js:398:26)
    at loader (/Users/montogeek/project/node_modules/babel-register/lib/node.js:126:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/montogeek/project/node_modules/babel-register/lib/node.js:136:7)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (currencies.js:2:1)
    at Module._compile (module.js:398:26)
    at loader (/Users/montogeek/project/node_modules/babel-register/lib/node.js:126:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/montogeek/project/node_modules/babel-register/lib/node.js:136:7)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (index.js:2:1)
    at Module._compile (module.js:398:26)
    at loader (/Users/montogeek/project/node_modules/babel-register/lib/node.js:126:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/montogeek/project/node_modules/babel-register/lib/node.js:136:7)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (index.js:5:1)
    at Module._compile (module.js:398:26)
    at loader (/Users/montogeek/project/node_modules/babel-register/lib/node.js:126:5)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/montogeek/project/node_modules/babel-register/lib/node.js:136:7)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at /Users/montogeek/project/node_modules/mocha/lib/mocha.js:219:27
    at Array.forEach (native)
    at Mocha.loadFiles (/Users/montogeek/project/node_modules/mocha/lib/mocha.js:216:14)
    at Mocha.run (/Users/montogeek/project/node_modules/mocha/lib/mocha.js:468:10)
    at Object.<anonymous> (/Users/montogeek/project/node_modules/mocha/bin/_mocha:403:18)
    at Module._compile (module.js:398:26)
    at Object.Module._extensions..js (module.js:405:10)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Function.Module.runMain (module.js:430:10)
    at startup (node.js:141:18)
    at node.js:980:3
```

When running:
`babel src --out-dir dist` it compiles without problem

**Comment by thgreasi** [comment]
@montogeek I'm not sure if this change has been published as part of a npm release, but #538 has completely removed the aforementioned code.
